Issue #23: Prevent race conditions

tomdickman opened the pull request:
Race conditions are evident as a result of running multiple
conversions in parallel through adhoc tasks, this results in one or
more conversions failing due to an access key being deleted for a
user when still in use.

Attach an instance ID to access keys and abstract the user key
handling into a key manager, so now only the key for a particular
filename is deleted, not all keys for user.

Closes #23

Commit:
Issue #23: Prevent race conditions

Race conditions are evident as a result of running multiple
conversions in parallel through adhoc tasks, this results in one or
more conversions failing due to an access key being deleted for a
user when still in use.

Attach an instance ID to access keys and abstract the user key
handling into a key manager, so now only the key for a particular
filename is deleted, not all keys for user.

dmitriim commented on classes/key_manager.php:
```diff
@@ -0,0 +1,144 @@
+ * @copyright  2021 Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class key_manager {
```
just wondering why did you decide to make all methods of this class static?

tomdickman replied:
So you don't have to instantiate it, there is no data stored in the class so it doesn't need to persist, it's just an API so I made it more of a functional class

dmitriim replied:
not all classes need to store data to be instantiated. I won't be picky here. So it's up to you to rewrite. Personally I'd prefer not to use static functions where possible.

dmitriim commented on classes/key_manager.php:
```diff
@@ -0,0 +1,144 @@
+     * @param string $key access key to use for user validation, this is required to login user and allow access of target page.
+     * @param int $instanceid the instance ID of key to login with.
+     */
+    final public static function login_with_key(string $key, int $instanceid) {
```
it feels like this method shouldn't be part of that class. I'd expect key_manager to deal with keys and login functionality could be moved to login_manager for example.

tomdickman replied:
Moved to a new `login_manager`

dmitriim commented on classes/key_manager.php:
```diff
@@ -0,0 +1,144 @@
+     *
+     * @throws \moodle_exception if the user ID was invalid.
+     */
+    protected static function setup_user_session(int $userid) {
```
this one should got to login_manage as well. Not dealing with keys

tomdickman replied:
Moved to `login_manager`

dmitriim commented on classes/key_manager.php:
```diff
@@ -0,0 +1,144 @@
+     *
+     * @return bool true on success.
+     */
+    public static function delete_user_key(int $instanceid): bool {
```
should it be delete_keys as it maybe potentially more than one?  Would be probably good to add userid as a parameter if we want to use it for a specific user.

tomdickman replied:
Renames and inject user ID rather than relying on global

dmitriim commented on classes/key_manager.php:
```diff
@@ -0,0 +1,144 @@
+        global $DB, $USER;
+
+        $record = [
+            'script' => 'tool/pdfpages',
```
a good practice to move this hardocded string to a class constant.

tomdickman replied:
Fixed

dmitriim commented on classes/key_manager.php:
```diff
@@ -0,0 +1,144 @@
+    /**
+     * Create a user key.
+     *
+     * @param int $instanceid the instance ID to create key for.
```
instanceid is a bit confusing. It seems like you are trying to match core create_user_key function, but there it is $instance and not $instanceid. What is it actually mean? Is it just a unique string?

tomdickman replied:
Renamed to `instance` in line with core APIs

dmitriim commented on classes/key_manager.php:
```diff
@@ -0,0 +1,144 @@
+     *
+     * @return int the instance ID.
+     */
+    public static function get_instance_id(string $filename): int {
```
it's more like generate_instance_id? Should we rename filename tp just $string?

tomdickman replied:
Renamed to `seed` as this is essentially what it is, a seed value for instance ID generation

dmitriim commented on classes/helper.php:
```diff
@@ -155,11 +131,18 @@ public static function get_pdf_filerecord(string $filename, string $converter) :
+            'instanceid' => $instanceid
+        ];
+
+        return new moodle_url('/admin/tool/pdfpages/index.php', $params);
```
how about to move this /admin/tool/pdfpages/index.php to a constant?

tomdickman replied:
Fixed

Commit:
Issue #23: Add login manager

- Move login related functions into dedicated login manager
- Rename 'instanceid' variables to 'instance' in line with core APIs

dmitriim commented on classes/converter.php:
```diff
@@ -78,9 +78,10 @@ final public function convert_moodle_url_to_pdf(moodle_url $url, string $filenam
-            $key = helper::create_user_key();
-            $proxyurl = helper::get_proxy_url($url, $key);
             $filename = ($filename === '') ? helper::get_moodle_url_pdf_filename($url) : $filename;
+            $instanceid = key_manager::get_instance_id($filename);
```
can we move this inside create_user_key? Basically we create a key for one URL only, right?  So knowing URL we can always recalculate instanceid. 

So basically we should rename our method key_manager::create_user_key to key_manager::create_user_key_for_url(string $url)

then inside we can do our encoding to save it to DB. 

 then  login_with_key  should get key and url to check if it's ok to use this key for that url like login_with_key(string $key, int $url) and then key_manager can check that.

In this case it's much more clear that we use key for url and not for abstract instanceid

tomdickman replied:
Good idea, I'll implement this

tomdickman replied:
Abstracted the instance generation and instead use the URL in `key_manager` and `login_manager` function parameters

dmitriim commented on classes/helper.php:
```diff
@@ -155,11 +131,18 @@ public static function get_pdf_filerecord(string $filename, string $converter) :
+        $params = [
+            'url' => $targeturl->out(),
+            'key' => $key,
+            'instanceid' => $instanceid
```
proxy don't need this as long if it knows URL. Then key manager can check by url

tomdickman replied:
Removed from proxy as abstracted to use URL

Commit:
Issue #23: Move hardcoded proxy URL into constant

Commit:
Issue #23: Move hardcoded script names into constant

Commit:
Issue #23: Rename function for deleting user keys

- Rename the function for deleting user keys to plural as it may be
misleading otherwise, since it deletes all user keys matching user
ID and instance
- Inject user ID rather than relying on global for creating and
deleting user keys

Commit:
Issue #23: Abstract access key instance, use URL instead

Abstract the access key unique instance generation used for
conversions and instead use the unique URL being converted.

Commit:
Issue #23: Fix incorrect spelling in phpdocs

Commit:
Issue #23: Pass user ID into URL access key creation
